### PR TITLE
Bxmsdoc 2288 Rules for Employee Rostering

### PR DESCRIPTION
**Pull request by gsheldon**
Requires SME and Peer review.

### CODE REVIEW

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc [hunk 1]
```diff
@@ -0,0 +1,32 @@
+[id='wb-employee-rostering-balance-employees-shift-rule-proc']
+= Create the `BalanceEmployeesShiftNumber` guided rule
```
Comment (yurloc) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: running the employee rostering sample project in decision central - mozilla firefox_042

This is hard to read. I'd prefer to remove the code-style formatting from within headings and also break the rule name down into individual words. This applies to all headings containing a rule name.

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc [hunk 2]
```diff
@@ -0,0 +1,32 @@
+[id='wb-employee-rostering-balance-employees-shift-rule-proc']
+= Create the `BalanceEmployeesShiftNumber` guided rule 
+
+The `BalanceEmployeesShiftNumber` guided rule creates a soft constraint that ensures all employees are assigned a shift, or that all employees are assigned an equal number of shifts. In the employee rostering example, this constraint is created using the guided rule editor.
```
Comment (yurloc) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: I think we should rephrase this to help the user better understand what the rule does. The issues I can see in the description

1. ensures all employees are assigned a shift - yes, but that's a side effect, not the main point of the rule
2. all employees are assigned an equal number of shifts - ideally yes, but that's not what the rule ensures

### How the rule works (WHEN section, condition)
For each employee it counts all his/her shift assignments. The number of shifts assigned to the employee is stored in the `$shiftCount` variable.
### How the rule works (THEN section, consequence)
Soft score is modified (penalized) with the negative value of the **square** of the shift count.

This is how the *rule* works and I'm not saying this needs to be explained. It's optional in my opinion. What's more important is to explain the *idea* of the constraint implemented by the rule. The idea is to penalize extreme shift counts by using the square as a score penalty. Example:

* 3 employees with 4, 5 and 6 shifts gives a score of -16 + -25 + -36 = -77 which is a worse total score than
* 3 employees with 5, 5 and 5 shifts, giving a total score of -25 + -25 + -25 = -75. This is a higher and therefore better soft score.

This example illustrates how the score formula, implemented by the rule, gives the solver an incentive to distribute shift assignments in a balanced way. If this were the only constraint, all employees would end up being assigned to equal number of shifts. Since, in a real world data set, this probably not possible due to other constraint, I would avoid stating that the rule "makes sure all employees are assigned an equal number of shifts". "The rule ensures that shifts are assigned in a more fair, balanced way" sounds more accurate.

@gsheldon the exact wording is up to you. I'm trying some extra details to help you understand how it works.

**Review by yurloc — CHANGES_REQUESTED** [review]

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc (hunk 2 above)
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: Search all instances of "editor" and change to "designer" when referring to "guided rules designer", "DRL designer", etc. Also things like "In the editor" should be "In the designer" etc. Per DM/BA terminology change. The only time should stay "editor" is when not referring to the designer (naturally), such as the **Expression editor** within the guided rules designer, the **Editor** tab (which will be changed in DM 7.1 anyway).

Here's the terminology change overview for DM/BA, if it's helpful: https://docs.google.com/spreadsheets/d/1kmK92hY3JJp7gZTvLzd0CawaZzvI2MYYv2uaLaQcMJc/edit#gid=1999150540

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+.Procedure
+. Go to *Menu* -> *Design* -> *Projects* and click the project name.
+. Click *Create New Asset* -> *Guided Rule*.
+. Enter `BalanceEmployeesShiftNumber` as the *Guided Rule* name and select the `employeerostering.employeerostering` *Package*. 
+. Click *Ok* to create the rule asset.
+. Add a *WHEN* condition by clicking the image:5686.png[] in the *WHEN* field.
+. Select `Employee` in the *Add a condition to the rule...* window. Click *+Ok*. 
+. Click on the `Employee` condition to modify the constraints and add the variable name `$employee`.
+. Add the *WHEN* condition `From Accumulate...`.
```
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: Per IBM SG, don't retain the trailing ellipsis when referring to UI text that has it at the end. So should be "the **Add a condition to the rule** window", "condition `From Accumulate`", etc. The only exception I've found is when the ellipsis is not the final element, such as "click **(show options...)**", at the bottom of the designer. I'd search all ellipses and see if they are per the UI and can be removed. Super minor, but just sayin'. Up to you.

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+. Click *Create New Asset* -> *Guided Rule*.
+. Enter `BalanceEmployeesShiftNumber` as the *Guided Rule* name and select the `employeerostering.employeerostering` *Package*. 
+. Click *Ok* to create the rule asset.
+. Add a *WHEN* condition by clicking the image:5686.png[] in the *WHEN* field.
+. Select `Employee` in the *Add a condition to the rule...* window. Click *+Ok*. 
+. Click on the `Employee` condition to modify the constraints and add the variable name `$employee`.
+. Add the *WHEN* condition `From Accumulate...`. 
+.. Above the `From Accumulate...` condition, click *click to add pattern...* and select `Number` as the fact type from the drop down list.
+.. Add the variable name `$shiftCount` to the `Number` condition.
+.. Below the `From Accumulate...` condition, click *click to add pattern...* and select the `ShiftAssignment` fact type from the drop down list.
```
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: drop-down list/button/etc. Ditto everywhere else this appears.

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+.. Add the variable name `$shiftAssignment` to the `ShiftAssignment` fact type.
+.. Click on the `ShiftAssignment` condition again and from the *Add a restriction on a field* drop down list, select `employee`.
+.. Select `equal to` from the drop down list next to the `employee` constraint.
+.. Click the image:6191.png[edit] icon next to the drop down button to add a variable, and click *Bound variable* in the *Field value* window.
+.. Select `$employee` from the drop down list.
+.. In the *Function* box type `count($shiftAssignment)`.
+. Add the *THEN* condition by clicking the image:5686.png[] in the *THEN* field.
+. Select `Modify Soft Score` in the *Add a new action...* window. Click *+Ok*.
+.. Type the following expression into the box: `-($shiftCount.intValue()*$shiftCount.intValue())`
+. Click *Validate* in the upper right corner to check all rule conditions are valid. If the rule validation fails, address any problems described in the error message, review all components in the rule, and try again to validate the rule until the rule passes.
```
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: upper-right corner/toolbar/etc. Ditto everywhere else this appears.

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+. Add the *WHEN* condition `From Accumulate...`. 
+.. Above the `From Accumulate...` condition, click *click to add pattern...* and select `Number` as the fact type from the drop down list.
+.. Add the variable name `$shiftCount` to the `Number` condition.
+.. Below the `From Accumulate...` condition, click *click to add pattern...* and select the `ShiftAssignment` fact type from the drop down list.
+.. Add the variable name `$shiftAssignment` to the `ShiftAssignment` fact type.
+.. Click on the `ShiftAssignment` condition again and from the *Add a restriction on a field* drop down list, select `employee`.
+.. Select `equal to` from the drop down list next to the `employee` constraint.
+.. Click the image:6191.png[edit] icon next to the drop down button to add a variable, and click *Bound variable* in the *Field value* window.
+.. Select `$employee` from the drop down list.
+.. In the *Function* box type `count($shiftAssignment)`.
```
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: Technically "type" should be "enter", per IBM SG (for clarity, supposedly). Another super minor thing. Completely your call. If you do change, be sure to update it everywhere else also.

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-shift-required-skills-met-proc.adoc [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+image::ShiftRequiredSkillsAreMet.png[ShiftRequiredSkillsAreMet Guided Rule]
+
+.Procedure
+. Go to *Menu* -> *Design* -> *Projects* and click the project name.
+. Click *Create New Asset* -> *Guided Rule*.
+. Enter `ShiftReqiredSkillsAreMet` as the *Guided Rule* name and select the `employeerostering.employeerostering` *Package*. 
+. Click *Ok* to create the rule asset.
+. Add a *WHEN* condition by clicking the image:5686.png[] in the *WHEN* field.
+. Select `ShiftAssignment` in the *Add a condition to the rule...* window. Click *+Ok*. 
+. Click on the `ShiftAssignment` condition, and select `employee` from the *Add a restriction on a field* drop down list.
```
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-shift-required-skills-met-proc.adoc]: Should typically avoid single numbered steps. Just make part of the actual step instead a single sub-step?

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-shift-required-skills-met-proc.adoc [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+.. In the editor, click the drop down list next to `employee` and select `is not null`.
+. Click on the `ShiftAssignment` condition, and click *Expression editor*.
+.. In the editor, click `[not bound]` to open the *Expression editor*, and bind the expression to the variable `$requiredSkill`. Click *Set*.
+.. In the editor, next to `$requiredSkill`, select `shift` from the first drop down list, then `requiredSkill` from the next drop down list.
+. Click on the `ShiftAssignment` condition, and click *Expression editor*.
+.. In the editor, next to `[not bound]`, select `employee` from the first drop down list, then `skills` from the next drop down list.
+.. Leave the next drop down list as `Choose...`.
+.. In the next drop down box, change `--please choose--` to `excludes`.
```
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-shift-required-skills-met-proc.adoc]: IMO, I think "change **--please choose--**" is in the same camp as final ellipses in the UI, which are to be deleted (per my previous comment). For that reason, I've just said "change **please choose**" and the like in my own docs, for whatever it's worth. Again, super minor and totally your call.

File: docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc (hunk 1 above)
Comment (sterobin) [inline docs/product-assembly_running_employee_rostering_in_wb/src/main/asciidoc/wb-employee-rostering-balance-employees-shift-rule-proc.adoc]: Agree, but not much you can do about it, given the CSS for H2's.

**Review by sterobin — COMMENTED** [review]
Here y' go, @gsheldon . Mostly small, inconsequential things around style and consistency. Nothing that will take long to address, if you choose to.

**Comment by gsheldon** [comment]
changes incorporated.
